README.md
# NAME

WorldCat::API - Fetch MARC::Records from OCLC's WorldCat API

# VERSION

version 1.002

# SYNOPSIS

```perl
my $api = WorldCat::API->new(
  institution_id         => "...",
  principle_id           => "...",
  principle_id_namespace => "...",
  secret                 => "...",
  wskey                  => "...",
);

my $marc_record = $api->find_by_oclc_number("123") or die "Not Found!";
```

# CONTRIBUTING

This project uses Dist::Zilla to set things up. You can run that directly or (more easily) by using Docker. For starters, create the build container:

```
$ docker build -t worldcatapi .
```

The container contains Perl, cpanm, dzil, and all of the build dependencies. Shell into it to use it as a dev environment:

```
$ docker run --volume="$PWD:/app" --entrypoint=/bin/bash worldcatapi
```

The "volume" flag syncs your local directory into the container, allowing you to develop interactively. That also means that if you build the app within the container, the build products will be reflected on your host machine:

```
$ docker run --volume="$PWD:/app" worldcatapi build
$ ls -l
…
WorldCat-API-1.002
WorldCat-API-1.002.tar.gz
…
```

# TESTING

To test, you must set (staging!) API credentials in the container environment. An easy solution is to add them to a .env file at the root of the project, which you can load with Docker:

```
$ cat <<EOF > .env
WORLDCAT_API_INSTITUTION_ID="..."
WORLDCAT_API_PRINCIPLE_ID="..."
WORLDCAT_API_PRINCIPLE_ID_NAMESPACE="..."
WORLDCAT_API_SECRET="..."
WORLDCAT_API_WSKEY="..."
EOF
$ docker run --volume="$PWD:/app" --env-file=.env worldcatapi test
```

# AUTHOR

Daniel Schmidt <danschmidt5189@gmail.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2018 by Daniel Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


changelog
0.001 Initial Commit
0.002 Package Cleanup
1.001 Fix descending version number error


lib/WorldCat/MARC/Record/Monkeypatch.pm
use strict;
use warnings;
package WorldCat::MARC::Record::Monkeypatch;

# ABSTRACT: Monkeypatch MARC::Record to add useful methods

=pod

=head1 SYNOPSIS

  use feature 'say';
  use MARC::Record;
  use WorldCat::MARC::Record::Monkeypatch;

  my $marc_record = MARC::Record->new_from_marc21xml($xmlobj);
  say $marc_record->oclc_number;

=cut

use feature qw(say);
use Time::Piece;

# Specs:
#   https://www.oclc.org/support/services/batchload/controlnumber.en.html
#
# NOTE(dcschmidt): Definitely incomplete; needs more tests/sample data.
sub MARC::Record::oclc_number {
  my ($self) = @_;

  if (my $f001 = $self->field('001')) {
    if ($f001->data =~ qr/^(ocm|ocn|on\.)(\d+)$/) {
      return int $2;
    }
  }

  if (my $f019 = $self->field('019')) {
    return int $f019->data;
  }

  if (my $f035a = $self->subfield('035', 'a')) {
    if ($f035a->data =~ qr/OCoLC.(\d+)$/) {
      return int $1;
    }
  }
}

sub MARC::Record::new_from_marc21xml {
  my ($class, $raw) = @_;

  my $marc_rec = "$class"->new;
  $marc_rec->leader($raw->{leader});

  # Add control fields
  for my $cfield (@{$raw->{controlfield}}) {
    $marc_rec->append_fields(MARC::Field->new(
      $cfield->{tag},
      $cfield->{content},
    ));
  }

  # Add data fields
  for my $dfield (@{$raw->{datafield}}) {
    my $content   = $dfield->{content};
    my $ind1      = $dfield->{ind1};
    my $ind2      = $dfield->{ind2};
    my $subfields = $dfield->{subfield};
    my $tag       = $dfield->{tag};

    # Normalize subfields to an array of hashrefs
    my %subfields = map { @{$_}{qw(code content)} } @{
      ref $subfields eq 'HASH' ? [$subfields] : $subfields
    };

    $marc_rec->append_fields(MARC::Field->new($tag, $ind1, $ind2, %subfields));
  }

  return $marc_rec;
}

1;


lib/WorldCat/API.pm
use strict;
use warnings;
package WorldCat::API;

# ABSTRACT: Fetch MARC::Records from OCLC's WorldCat API

=pod

=head1 SYNOPSIS

  my $api = WorldCat::API->new(
    institution_id         => "...",
    principle_id           => "...",
    principle_id_namespace => "...",
    secret                 => "...",
    wskey                  => "...",
  );

  my $marc_record = $api->find_by_oclc_number("123") or die "Not Found!";

=head1 CONTRIBUTING

This project uses Dist::Zilla to set things up. You can run that directly or (more easily) by using Docker. For starters, create the build container:

  $ docker build -t worldcatapi .

The container contains Perl, cpanm, dzil, and all of the build dependencies. Shell into it to use it as a dev environment:

  $ docker run --volume="$PWD:/app" --entrypoint=/bin/bash worldcatapi

The "volume" flag syncs your local directory into the container, allowing you to develop interactively. That also means that if you build the app within the container, the build products will be reflected on your host machine:

  $ docker run --volume="$PWD:/app" worldcatapi build
  $ ls -l
  …
  WorldCat-API-1.002
  WorldCat-API-1.002.tar.gz
  …

=head1 TESTING

To test, you must set (staging!) API credentials in the container environment. An easy solution is to add them to a .env file at the root of the project, which you can load with Docker:

  $ cat <<EOF > .env
  WORLDCAT_API_INSTITUTION_ID="..."
  WORLDCAT_API_PRINCIPLE_ID="..."
  WORLDCAT_API_PRINCIPLE_ID_NAMESPACE="..."
  WORLDCAT_API_SECRET="..."
  WORLDCAT_API_WSKEY="..."
  EOF
  $ docker run --volume="$PWD:/app" --env-file=.env worldcatapi test

=cut

use feature qw(say);

use Moo;
use Carp qw(croak);
use Digest::SHA qw(hmac_sha256_base64);
use HTTP::Request;
use HTTP::Status qw(:constants);
use LWP::UserAgent;
use MARC::Record;
use Math::Random::Secure qw(irand);
use Readonly;
use WorldCat::MARC::Record::Monkeypatch;
use XML::Simple qw(XMLin);

Readonly my $DEFAULT_RETRIES => 5;

sub _from_env {
  my ($attr) = @_;
  return $ENV{uc "WORLDCAT_API_$attr"} // die "Attribute $attr is required";
}

has institution_id => (
  is => 'ro',
  required => 1,
  default => sub { _from_env('institution_id') },
);

has principle_id => (
  is => 'ro',
  required => 1,
  default => sub { _from_env('principle_id') },
);

has principle_id_namespace => (
  is => 'ro',
  required => 1,
  default => sub { _from_env('principle_id_namespace') },
);

has secret => (
  is => 'ro',
  required => 1,
  default => sub { _from_env('secret') },
);

has wskey => (
  is => 'ro',
  required => 1,
  default => sub { _from_env('wskey') },
);

sub _query_param {
  return "$_[0]=\"$_[1]\"";
}

# OCLC returns encoding=UTF-8, format=MARC21+xml.
sub find_by_oclc_number {
  my ($self, $oclc_number, %opts) = @_;

  my $retries = $opts{retries} // $DEFAULT_RETRIES;

  # Fetch the record with retries and exponential backoff
  my $res;
  my $ua = $self->_new_ua;
  for my $try (0..($retries - 1)) {
    $res = $ua->get("https://worldcat.org/bib/data/$oclc_number");
    say "Got HTTP Response Code: @{[$res->code]}";

    last if not $res->is_server_error; # only retry 5xx errors
    sleep 2 ** $try;
  }

  # Return MARC::Record on success
  if ($res->is_success) {
    my $xml = XMLin($res->decoded_content)->{entry}{content}{record};
    return MARC::Record->new_from_marc21xml($xml);
  }

  # Return nil if record not found
  return if $res->code eq HTTP_NOT_FOUND;

  # An error occurred, throw the response
  croak $res;
}

# Generate the authorization header. It's complicated; see the docs:
#
#   https://www.oclc.org/developer/develop/authentication/hmac-signature.en.html
#   https://github.com/geocolumbus/hmac-language-examples/blob/master/perl/hmacAuthenticationExample.pl
sub _create_auth_header {
  my ($self) = @_;

  my $signature = $self->_create_signature;

  return 'http://www.worldcat.org/wskey/v2/hmac/v1 ' . join(q{,},
    _query_param(clientId      => $self->wskey),
    _query_param(principalID   => $self->principle_id),
    _query_param(principalIDNS => $self->principle_id_namespace),
    _query_param(nonce         => $signature->{nonce}),
    _query_param(signature     => $signature->{value}),
    _query_param(timestamp     => $signature->{timestamp}),
  );
}

sub _create_signature {
  my ($self, %opts) = @_;

  my $nonce = $opts{nonce} || sprintf q{%x}, irand;
  my $timestamp = $opts{timestamp} || time;

  my $signature = hmac_sha256_base64(join(qq{\n},
    $self->wskey,
    $timestamp,
    $nonce,
    q{}, # Hash of the body; empty because we're just GET-ing
    "GET", # all-caps HTTP request method
    "www.oclc.org",
    "443",
    "/wskey",
    q{}, # query params
  ), $self->secret) . q{=};

  return {
    value     => $signature,
    nonce     => $nonce,
    timestamp => $timestamp,
  };
}

sub _new_ua {
  my ($self) = @_;

  my $ua = LWP::UserAgent->new;
  $ua->default_header(Accept => q{application/atom+xml;content="application/vnd.oclc.marc21+xml"});
  $ua->default_header(Authorization => $self->_create_auth_header);
  return $ua;
}

1;
